tracer/src/Datadog.Trace/Sampling/SpanSampler.cs

```diff
@@ -0,0 +1,58 @@
+// <copyright file="SpanSampler.cs" company="Datadog">
+// Unless explicitly stated otherwise all files in this repository are licensed under the Apache 2 License.
+// This product includes software developed at Datadog (https://www.datadoghq.com/). Copyright 2017 Datadog, Inc.
+// </copyright>
+
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using Datadog.Trace.Logging;
+
+namespace Datadog.Trace.Sampling;
+
+/// <summary>
+///     Represents a sampler for single span ingestion.
+/// </summary>
+internal class SpanSampler : ISpanSampler
+{
+    private static readonly IDatadogLogger Log = DatadogLogging.GetLoggerFor<SpanSampler>();
+
+    private readonly List<ISpanSamplingRule> _rules = new List<ISpanSamplingRule>();
+
+    public SpanSampler(IEnumerable<ISpanSamplingRule> rules)
+    {
+        _rules = rules.ToList() ?? throw new ArgumentNullException(nameof(rules));
+    }
+
+    /// <inheritdoc/>
+    public void MakeSamplingDecision(Span span)
+    {
+        if (_rules.Count > 0)
+        {
+            foreach (var rule in _rules)
+            {
+                if (rule.ShouldKeep(span))
+                {
+                    Tag(span, rule);
+                    return;
+                }
+            }
+        }
+    }
+
+    /// <inheritdoc/>
+    public void Tag(Span span, ISpanSamplingRule rule)
+    {
+        // TODO do we want to tag here?
+        // TODO maybe do a TryAdd? is SetTag safe?
+        span.SetTag(Tags.SingleSpanSampling.RuleRate, rule.SamplingRate.ToString());
```

Comment:
Premature optimization: You may want to cache the 3 string you use here so you don't keep allocating them every time you add the tags. Maybe cache them in rule itself. Something like:
```csharp
// pseudo-code, in the class that implements ISpanSamplingRule
SamplingRateString { get; } = SamplingRate.ToString();
MaxPerSecondString { get; } = MaxPerSecond.ToString();
SpanSamplingRuleString { get; } = SpanSamplingRule.ToString();
```

Just brainstorming, this may not be the best way to cache them, but we try to be vigilant about any heap allocations.